### ISSUE TITLE
MenuItem: CommandParameters are ignored

### ISSUE THREAD

**Issue by SetTrend** [dotnet/wpf#4078]
- .NET Core Version: 5
- Windows version: 20H2

## Problem description:

I created an MVVM sample repository. In this repository I'm loading [`MenuItem`](https://docs.microsoft.com/dotnet/api/system.windows.controls.menuitem)s and [`Button`](https://docs.microsoft.com/dotnet/api/system.windows.controls.button)s with commands.

While the `Button` elements call `ICommand::Execute(object? parameter)` with the given `CommandParameter`, `MenuItem` elements don't.

Here's a code snipped showing both declarations:

```xml
<MenuItem Header="Film_liste" ItemsSource="{Binding Path=Commands}">
    <MenuItem.ItemTemplate>
      <DataTemplate>
        <MenuItem Command="{Binding}" CommandParameter="{Binding ElementName=Filmliste, Path=SelectedItem}" Header="{Binding Path=Text}" />
      </DataTemplate>
    </MenuItem.ItemTemplate>
  </MenuItem>

  ...

  <ItemsControl.ItemTemplate>
    <DataTemplate>
      <Button Command="{Binding}" CommandParameter="{Binding ElementName=Filmliste, Path=SelectedItem}" Content="{Binding Path=Text}" Margin="5" Padding="10 3"></Button>
    </DataTemplate>
  </ItemsControl.ItemTemplate>
```
 
## Actual behavior:

Commands triggered by `MenuItem` elements are always passed `null` as command parameter.
 
## Expected behavior:

Commands triggered by `MenuItem` elements should be passed the value provided as command parameter.
 
## Minimal repro:

[MVVM Test repository](/SetTrend/MVVM-Test)

## Steps to reproduce:

https://user-images.githubusercontent.com/9283914/105822923-da8b8880-5fbc-11eb-9c90-e665297a710a.mp4

**Comment by miloush** [comment]
Your `MenuItem.ItemTemplate` is not doing what you think it is doing - you are creating a `MenuItem` inside a `MenuItem`. The binding for the inner one works as expected, however, there is no binding on the outer one so it gets disabled which applies to its children too.

Try this instead:
```
<MenuItem Header="Film_liste" ItemsSource="{Binding Path=Commands}">
    <MenuItem.ItemContainerStyle>
        <Style TargetType="MenuItem">
            <Setter Property="Command" Value="{Binding}" />
            <Setter Property="CommandParameter" Value="{Binding ElementName=Filmliste, Path=SelectedItem}" />
            <Setter Property="Header" Value="{Binding Path=Text}" />
        </Style>
    </MenuItem.ItemContainerStyle>
</MenuItem>
```

**Comment by SetTrend** [comment]
This is strange ...

According to the book ["Pro WPF 4.0 in C#" from Apress](https://www.apress.com/gp/book/9781430243656) and [Microsoft docs](https://docs.microsoft.com/archive/msdn-magazine/2009/february/patterns-wpf-apps-with-the-model-view-viewmodel-design-pattern#mainwindowviewmodel-class) this should work:

```c#
<!-- In MainWindow.xaml -->
<Menu>
    <MenuItem Header="_File">
        <MenuItem Header="_Exit" 
                    Command="{Binding Path=CloseCommand}" />
    </MenuItem>
    <MenuItem Header="_Edit" />
    <MenuItem Header="_Options" />
    <MenuItem Header="_Help" />
</Menu>
```

The outer `MenuItem` doesn't have a binding here, too.

**Comment by SetTrend** [comment]
I think it's slowly getting in my grasp ...

So you suggest, the structure created from my original code is like:

```xml
<Menu DockPanel.Dock="Top">
  <MenuItem Header="Film_liste" ItemsSource="{Binding Path=Commands}">
    <MenuItem>
      <MenuItem Command="{Binding}" CommandParameter="{Binding ElementName=Filmliste, Path=SelectedItem}" Header="{Binding Path=Text}" />
```

Shouldn't this layout actually be presented in the Visual Live Tree:


![WPF MenuItems](https://user-images.githubusercontent.com/9283914/107131036-74173c00-68d3-11eb-8a66-ce16bf7d7fd9.png)

**Comment by miloush** [comment]
No, I was suggesting your code was putting MenuItem sort of inside as it was the Header of the container-MenuItem.

Your book example, on the other hand, is creating submenus.

**Comment by SetTrend** [comment]
Did you read my second comment?

**Comment by SetTrend** [comment]
I believe I can prove a bug in the way commands in `MenuItem`s are refreshed.

`MenuItem`s are calling `CanExecute` when the main menu is opened. However, the `CommandParameter` provided to `CanExecute`  isn't reliably acquiring the data bound to them.
<br/>

### In my example project, I can reliably reproduce the erroneous behaviour:

When you select an item from the `ListBox`, then open the main menu for the first time, the `CommandParameter` binding is not getting performed. **ERROR**

If, however, you open the main menu first, then select an item from the `ListBox` and then open the main menu again, binding is getting performed correctly.
<br/>

### Here's the flow chart, reporting the two approaches:

![Flow chart](https://user-images.githubusercontent.com/9283914/107966653-756bf700-6fac-11eb-9fa4-55f47d3ad22e.png)
<br/>

### Here's a screencast, reporting a debug session performing the successful path:

![MenuItem doesn't convey CommandParameter - 2](https://user-images.githubusercontent.com/9283914/107967461-6c2f5a00-6fad-11eb-8864-d0ee0a95cf05.gif)
<br/>

### And here's a screencast, reporting a debug session performing the error path:

![MenuItem doesn't convey CommandParameter - 3](https://user-images.githubusercontent.com/9283914/107967484-73566800-6fad-11eb-8dac-e00b11538dc3.gif)
<br/>

### In case anyone needs to manually pause/rewind, here's the original screencast recording:

https://user-images.githubusercontent.com/9283914/107966675-7ef55f00-6fac-11eb-89ae-16c8b7068710.mp4

**Comment by SetTrend** [comment]
Because this thread has grown old an possibly out of sight, I'd like to invite @ryalanms to this bug report.

**Comment by miloush** [comment]
Then this seems to be a duplicate of #3452.

(changing command parameter does not actually re-evaluate CanExecute)

**Comment by SetTrend** [comment]
Are you sure? After reading the other issue I don't think so, because:

- in my case, `CanExecute` _is_ getting called
- it's just the _parameter_ that's provided to `CanExecute` that isn't evaluated/provided correctly

So, in my case the `CommandParameter` binding isn't reliably evaluated before `CanExecute` is getting called.

Plus, in contrast to the other case, in my case the issue seems to only exist for `MenuItem` components. The same code in the same WPF form works flawlessly for `Button` components. (See my minimum sample repository referenced above and my screencast.)

**Comment by SetTrend** [comment]
I suppose, someone with sufficient Member's/Contributor's expertise, equipped with a debug build of WPF Core should smoothly be able to find and fix the binding issue I reported by debugging into the WPF framework function that's calling `CanExecute` for the menu item.

I provided all that's necessary to reproduce the issue: A minimum MVVM WPF project and detailed steps how to reproduce.

**Comment by miloush** [comment]
I'm sorry you are stuck with me. I am pretty sure, you can see yourself be setting the command parameter first:

Change
```xml
        <Style TargetType="MenuItem">
            <Setter Property="Command" Value="{Binding}" />
            <Setter Property="CommandParameter" Value="{Binding ElementName=Filmliste, Path=SelectedItem}" />
            <Setter Property="Header" Value="{Binding Path=Text}" />
        </Style>
```

to

```xml
        <Style TargetType="MenuItem">
            <Setter Property="CommandParameter" Value="{Binding ElementName=Filmliste, Path=SelectedItem}" />
            <Setter Property="Command" Value="{Binding}" />
            <Setter Property="Header" Value="{Binding Path=Text}" />
        </Style>
```

**Comment by SetTrend** [comment]
Sincerely pardon me, @miloush, no insulting intended when I wrote my previous comment. I'm sorry if it inadvertently caused irritation on your behalf.
<br/>

After reading your comment, I found the following blogs about this issue:

- [WPF: XAML Attribute Order Matters](https://www.nickdarnell.com/xaml-attribute-order-matters/)
- [The Order Issue of XAML Attributes](https://weblogs.asp.net/dixin/the-order-issue-of-xaml-attributes)
<br/>

I don't think, this exotic behaviour is documented. I can't find any documentation on this. On the contrary - The following is the current state of WPF attribute sequence documentation I'm aware of:
<br/>

#### Matthew MacDonald, "[Pro WPF 4.5 in C#](https://www.apress.com/gp/book/9781430243656)", 4<sup>th</sup> edition, Apress

Page 102, "The Coercion Callback"

> This is an important consideration for initialization, such as when a window is being created for a XAML document. All WPF controls guarantee that their properties can be set in any order, without causing any change in behavior.
> [...]
>This behavior is due to WPF’s property resolution system, which you learned about earlier. Although WPF stores the exact local value you’ve set internally, it evaluates what the property should be (using coercion and a few other considerations) when you read the property.
<br/>

#### Microsoft Docs, "[Coerce Value Callbacks and Property Changed Events](https://docs.microsoft.com/dotnet/desktop/wpf/advanced/dependency-property-callbacks-and-validation#coerce-value-callbacks-and-property-changed-events)"

> Using these two callbacks in combination, you can create a series of properties on elements where changes in one property will force a coercion or reevaluation of another property.
<br/>

#### Microsoft Docs, "[DependsOnAttribute Class](https://docs.microsoft.com/dotnet/api/system.windows.markup.dependsonattribute)"

>Indicates that the attributed property is dependent on the value of another property.
> [...]
> The [`Save`](https://docs.microsoft.com/dotnet/api/system.windows.markup.xamlwriter.save) method on [`XamlWriter`](https://docs.microsoft.com/dotnet/api/system.windows.markup.xamlwriter) will process the specified property before processing the property that this attribute is set on.

However:

##### Microsoft Docs, "[MenuItem.CommandParameter Property](https://docs.microsoft.com/dotnet/api/system.windows.controls.menuitem.commandparameter)"

```c#
[System.ComponentModel.Bindable(true)]
[System.Windows.Localizability(System.Windows.LocalizationCategory.NeverLocalize)]
public object CommandParameter { get; set; }
```
<hr/>

If this behaviour is not documented ...

- will it cause issues with the Visual Studio WPF designer?
- will it cause issues with the [`Save`](https://docs.microsoft.com/dotnet/api/system.windows.markup.xamlwriter.save) method on [`XamlWriter`](https://docs.microsoft.com/dotnet/api/system.windows.markup.xamlwriter)
<br/>

I can test and implement your suggested workaround. Then I will have a local solution and have learned from life. - However, **shouldn't this be fixed in WPF**, once and for all?

**Comment by miloush** [comment]
So what happens is that when you open the menu for the first time, the MenuItems get created - the constructor is called, and all the properties are set in the order that it is specified in XAML. That way the `MenuItem.Command` is bound and set to a value, which, due to it's property changed callback, reevaluates the command's `CanExecute` status:

See the `OnCommandChanged` > `OnCommandChanged` > `HookCommand` > `UpdateCanExecute` sequence here:
https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/MenuItem.cs,407

Note that at this point, the `CommandParameter` has not been set (bound) yet, and therefore `CanExecute` receives `null` parameter.

Now the loader sets (bounds) the `CommandParameter`, but there is no property changed handler for that, so it does not re-evaluate it.

Now when you are done loading, both `Command` and `CommandParameter` are bound and will stay current thanks to the databinding. Once the item is loaded, any operations, including `Save` will therefore be fine.

> All WPF controls guarantee that their properties can be set in any order, without causing any change in behavior.

I don't see such guarantee in the documentation, and there are other types where the order matters (e.g. whether you set first `WindowStyle` or `WindowState` determines whether taskbar will be covered by a maximized borderless window).

I do agree, however, that pretending that `CommandParameter` does not affect whether `Command` can be executed is a design overview. I believe that is covered by #3452. If changing the `CommandParameter` would invalidate the `CanExecute`, then it will be called twice during loading, once with `null` parameter and one with the set one. That I believe is the correct behaviour

**Comment by SetTrend** [comment]
Shouldn't [this line in `MenuItem.OpenMenu`](https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/MenuItem.cs,2386):

```c#
if ((owner != null) && ((owner is MenuItem) || (owner is MenuBase)))
  {
    // Parent must be MenuItem or MenuBase in order for menus to open.
    // Otherwise, odd behavior will occur.
    SetCurrentValueInternal(IsSubmenuOpenProperty, BooleanBoxes.TrueBox);
    return true; // The value was actually changed
  }
```

... be rather amended by iterating through the ` MenuItem`'s children to call their `UpdateCanExecute` method, similar to this:

```c#
if ((owner != null) && ((owner is MenuItem) || (owner is MenuBase)))
  {
    // Parent must be MenuItem or MenuBase in order for menus to open.
    // Otherwise, odd behavior will occur.
    SetCurrentValueInternal(IsSubmenuOpenProperty, BooleanBoxes.TrueBox);

    foreach (MenuItem mi in Children) mi.UpdateCanExecute();

    return true; // The value was actually changed
  }
```

**Comment by miloush** [comment]
#316 - this is the issue I was looking for

**Comment by SetTrend** [comment]
We've been writing in parallel. Did you read my last comment? I'll be reading your references now.

**Comment by miloush** [comment]
I don't think you should invalidate all `CanExecute` for each item when nothing changed, that might have non-trivial performance costs.

**Comment by SetTrend** [comment]
I believe you need to. Because only at the moment when a menu opens you get the current program state, particularly when the `CommandParameter` is a binding expression. That's something you cannot cache.

**Comment by miloush** [comment]
What scenario do you have in mind where calling `CanExecute` after `CommandParameter` is changed would not fix the issue but re-evaluating on menu opening would? I don't see why opening a menu should be a particularly important event to achieve correctness, since menus can be "always open" (for example the top level menu). We should aim for the current state being always reflected, not only at specific points in time.

**Comment by SetTrend** [comment]
Ah, OK, I see. Right ...

The issue is with `CommandParameter` being a `Binding` expression. In that event, `CommandParameter` never changes, but the actual value bound to `CommandParameter` changes.

So, is it possible to bind an event to when the actual value bound to `CommandParameter` changes (in a binding expression that'd be the value that's bound)? Wouldn't this be beneficial?

**Comment by SetTrend** [comment]
Something similar to (I'm writing a coarse idea here):

```c#
[CommonDependencyProperty]
public static readonly DependencyProperty CommandParameterProperty = DependencyProperty.Register( "CommandParameter", typeof(object), typeof(...), new FrameworkPropertyMetadata((object) null),
  param => switch (param)
  {
    case Binding b:
      b.TargetUpdated += UpdateCanExecute();
  }
);
```

**Comment by miloush** [comment]
Yes, that's exactly what #3452 is asking for. Sorry I didn't make it clear, the property changed callback is called anytime the value is changed, which can come from binding source value change, style change, direct value assignment etc., (i.e. it does not only mean "assigning a binding to it")

This is how it's done with the `Command` property:

https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/MenuItem.cs,389

You want the same with `CommandParameter` where the `OnCommandParameterChanged` would call `UpdateCanExecute()`.

**Comment by SetTrend** [comment]
Ahh ... now I get it!

Perfect. So, I voted for that issue now.

Shall we close this issue here then? Or keep it open for documentation reasons?

You've been great, very helpful, thorough, thoughtful. Thanks so much for your patience in enlightening me!

**Comment by miloush** [comment]
You're welcome, sorry for the delays in responding.

I think you can close this issue (as duplicate of #316), it will stay available for documentation purposes.

**Comment by SetTrend** [comment]
@miloush:

Just one more question:

I can see that issues #3452 and #316 don't reference each other. Wouldn't it be beneficial to mention one from the other so members of each thread get aware of the other thread?

**Comment by miloush** [comment]
Sure, go ahead!

**Comment by GF-Huang** [comment]
So what's the solution?
